File: src/main/java/lab6/string_list/ListReverseIterator.java
package lab6.string_list;

import java.util.NoSuchElementException;

public class ListReverseIterator extends ListIterator {
    ListReverseIterator(Node current, StringList list) {
        super(current, list);
    }

    @Override
    public boolean hasNext() {
        return current != null;
    }

    @Override
    public String next() {
        Node prev = current;
        if (prev == null) {
            throw new NoSuchElementException("Iterator doesn't point to prev element!");
        }
        current = current.getPrev();
        return prev.getValue();
    }
}

File: src/main/java/lab7/my_list/ListReverseIterator.kt
package lab7.my_list

class ListReverseIterator<T>(current: Node<T>? = null, array: MyList<T>) : ListIterator<T>(current, array) {
    override operator fun hasNext(): Boolean {
        return current != null
    }

    override operator fun next(): T {
        val prev = current ?: throw NoSuchElementException("Iterator doesn't point to next element!")
        current = current!!.getPrev()
        return prev.getValue()!!
    }
}